Use helpers in Guid's ROS<byte> constructor

DaZombieKiller opened the pull request:
Closes https://github.com/dotnet/runtime/issues/78440. ([SharpLab demo](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgEllfCAAcAynwBugsDFzTqNeTGwATAPLCANuP27swgDzBxGGAB8DADiMBjBHIJ2bH7mABQAlAwAvEHCMADuDL7+ANoAugDedAh05GgMJXSkFVUAzLWl6JWlAKyNdEgdAOwdABwdAJwd2B3AHWAddh0wHVyN5HQAvpY0EVEMALLkiSlpmSGRdnGh4Ucx/riJCavU63akW6S7qQzpWffPp/cX8Qk3NBouX0GCgHDAGAAMthxBAOBg4tDYfCANIiOwsQwARw4MGUgmwzgS+RosHsEBc4gYuFB4IwhyipBohRoDDZDF0UEERmw/gYZLsFNcDBE9IA+thLOyOVyeXyBUKqbgABbQcXAKXsznc3kwfm2QWU6mqqDisCatnauV6hVGnJ6sV2C0ynXyg2K7KxBhimDOq26/Xku1esVcP2ygO24X270Ac3Drpt7uDfLFyoT1sDhujIcEGcjyZzqYAVvm3UGiw6ANbO1ns4h1BkPOKihijY1q7IVFWdyaevnTft62ZDhjzUexiox5VTr2CWd84sLvVVhJ1tks6jS6USlJt5074B7jXr7disB782nnd2PdO6/sn1730Ptmhvdh19xvfxr9pvfpn+gh7nmf7FnupZ/lWe41qeSyAlu9aNp8cQ2PYTiuO4ng+LEQTAGuiEbl+ghcAwcTACwkJ4rGGDKgwACEyQMOQSAEdu7Kbux7EACrKlAEAZAAglAsYcGoygAKIIGYugqBSiQHuy8FWIR24kWRbCCBgADCFJGHw/hQJouCQlpGDODAEnCHYBLCGxXGcVx0q0YIuB7psMD4NA4ibGIKqEiwaF2F4nwBORAKqU5xDdIpbLKV+LluUxQX3KZGDmZZ1m2eFsUMF+zBIEwKAMLx/FCSJYl4hgUkyXJwiJF+jlOWytFlW8BzCaJ4nVdJMCyYI8kAES/G2UBQDCY7QE2DD4BwNLZMOCDYBCwosUObnOBSsYsINFSiGoEBcOFEVOfFkX1mtnwMClRxpRlVk2Z4qEGhhbgeN49p4fZ7FNc1MbQUxwC5OQrT5Ll0pftK0XtRkcSQ1xmmiBIAAKXL4Fp3LmIFBpaBgdSkHdFkPdl+FoPD7GI2I4io6oGP6QIQW4yxhOZY99UUfozimDAcQoP8ZPnc1DCUyjaN01jjPKMzZlE1lT0c1zZhxKxCQC0L7FA30+Rq+r0pA4M2vk9uQOLIbgt/cD5Bm7revA6Q1s22yJt1A7jsmygrs21WRsnVxynSspSxAA))
Moves the big endian path and exception throwing to a helper, resulting in improved codegen when the constructor is inlined.

DaZombieKiller committed:
Use helpers in Guid's ROS<byte> constructor

EgorBo commented:
Technically, the BigEndian path doesn't need a helper, it's trimmed out anyway. It's the throw helper that messes inlining up:

```
*************** In fgFindBasicBlocks() for System.Guid:.ctor(System.ReadOnlySpan`1[ubyte]):this
weight= 77 : state  16 [ ldarga.s ]
weight= 79 : state  40 [ call ]
weight= 41 : state  32 [ ldc.i4.s ]
weight=  6 : state  46 [ beq.s ]
weight= 79 : state  40 [ call ]
weight= 66 : state 102 [ ldstr ]
weight= 79 : state  40 [ call ]
weight= 66 : state 102 [ ldstr ]
weight=227 : state 103 [ newobj ]
weight=210 : state 108 [ throw ]
weight=159 : state 112 [ ldsfld ]
weight=-24 : state  39 [ pop ]
weight= 10 : state   3 [ ldarg.0 ]
weight= 16 : state   4 [ ldarg.1 ]
weight= 79 : state  40 [ call ]
weight= 36 : state 115 [ stobj ]
weight= 19 : state  42 [ ret ]

multiplier in instance constructors increased to 1.5.
multiplier in methods of struct increased to 4.5.
1 arguments are structs passed by value.  Multiplier increased to 6.5.
Inline candidate has an arg that feeds a constant test.  Multiplier increased to 7.5.
Inline candidate callsite is boring.  Multiplier increased to 8.8.
calleeNativeSizeEstimate=1225
callsiteNativeSizeEstimate=135
benefit multiplier=8.8
threshold=1188
Native estimate for function size exceeds threshold for inlining 122.5 > 118.8 (multiplier = 8.8)
```

tannergooding commented:
> Technically, the BigEndian path doesn't need a helper, it's trimmed out anyway. 

Is it properly trimmed out? The linker has a bad habit of leaving comparisons and branches that do nothing and that likely negatively impacts some analysis and opts we can do

EgorBo commented:
> > Technically, the BigEndian path doesn't need a helper, it's trimmed out anyway.
> 
> Is it properly trimmed out? The linker has a bad habit of leaving comparisons and branches that do nothing and that likely negatively impacts some analysis and opts we can do

I've attached the IL ^, looks pretty normal to me 🤷

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -50,7 +50,7 @@ public Guid(ReadOnlySpan<byte> b)
+                ThrowArgumentException();
             }
 
             if (BitConverter.IsLittleEndian)
```
This can just be:
```C#
this = BitConverter.IsLittleEndian ? MemoryMarshal.Read<Guid>(b) : ReadGuidLittleEndian(b);
```

DaZombieKiller committed:
Condense Guid ROS<byte> constructor with a ternary

stephentoub commented:
Thanks.  I just read @EgorBo's comment.  Given that, I think we should just outline the throw and keep the rest the same.  Otherwise, we're possibly regressing Guid's ctor on big endian for little reason.

DaZombieKiller committed:
Remove ReadGuidLittleEndian helper

DaZombieKiller commented:
That makes sense. When I added the `ReadGuidLittleEndian` helper I was in the "portable assembly" mindset and forgot that the runtime is always trimmed, rendering the helper redundant. I've removed it in https://github.com/dotnet/runtime/pull/78446/commits/65763acf9902a414c0a36da1cb9f81088ab8d2eb.

stephentoub reviewed (approved):
Thanks.

MichalPetryka commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -71,6 +71,11 @@ public Guid(ReadOnlySpan<byte> b)
             _i = b[13];
             _j = b[14];
+
+            static void ThrowArgumentException()
```
```suggestion
            [StackTraceHidden]
            static void ThrowArgumentException()
```
in order to keep the previous stacktrace.

EgorBo replied:
Previously we'd see the following stack trace in case of an exception:
```
Guid..ctor(Span)
MethodCallingGuid
```
With @DaZombieKiller's change it's now:
```
Guid.ThrowArgumentException
MethodCallingGuid
```
with the attribute you suggest, the stack trace will be:
```
MethodCallingGuid
```

DaZombieKiller replied:
Addressed in https://github.com/dotnet/runtime/pull/78446/commits/057f68cd888dca03b405fb1f9d7bb93c4c241963.

danmoseley replied:
Why this particular throw helper and not others?

xtqqczze commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -60,7 +60,7 @@ public Guid(ReadOnlySpan<byte> b)
 
             // slower path for BigEndian:
-            _k = b[15];  // hoist bounds checks
+            _k = b[15]; // hoist bounds checks
```
nit: revert the whitespace change

DaZombieKiller replied:
That must have been an accident on my part, I didn't notice the double space in the original code. Addressed in https://github.com/dotnet/runtime/pull/78446/commits/8c2907ebb3d3988e01d00ece3060c406e86739f2.

DaZombieKiller committed:
Add [StackTraceHidden] to Guid.ThrowArgumentException

DaZombieKiller committed:
Revert accidental whitespace change

rickbrew commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -71,6 +71,12 @@ public Guid(ReadOnlySpan<byte> b)
             _j = b[14];
+
+            [StackTraceHidden]
+            static void ThrowArgumentException()
```
Should this also have `[DoesNotReturn]` ?

stephentoub replied:
It _could_, but it wouldn't help anything here.  [DoesNotReturn] was introduced as part of nullable reference types, and let's the compiler factor into its nullability flow analysis that the method won't return.  There's only one call site here and there aren't reference types involved.
